### PR TITLE
Add new option for auto importing names

### PR DESCRIPTION
**Pull request by sabbelasichon**
- In order to apply the Option::AUTO_IMPORT_NAMES only on those files which are really changed by a rector we introduce this new option

### CODE REVIEW

**Comment by TomasVotruba** [comment]
I wonder if this should be default behavior. What do you think?

**Comment by sabbelasichon** [comment]
> I wonder if this should be default behavior. What do you think?

I have people which are super happy to use these kind of rector by itself. Maybe we should extract this kind of rector to a real rector and only use the NameImportingPostRector for the files touched by rector itself. Hope my explanation is clear

**Comment by samsonasik** [comment]
@TomasVotruba the user may run rector only for auto import verify, I think it should be optional

**Comment by TomasVotruba** [comment]
> I have people which are super happy to use these kind of rector by itself. Maybe we should extract this kind of rector to a real rector and only use the NameImportingPostRector for the files touched by rector itself. Hope my explanation is clear

I understand, that makes lot of sense actually :+1:

File: rules-tests/CodingStyle/Rector/Namespace_/ImportFullyQualifiedNamesRector/AutoImportNamesOnlyOnChangedFiles.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+namespace Rector\Tests\CodingStyle\Rector\Namespace_\ImportFullyQualifiedNamesRector;
+
+use Iterator;
+use Rector\Testing\PHPUnit\AbstractRectorTestCase;
+use Symplify\SmartFileSystem\SmartFileInfo;
+
+/**
+ * @see \Rector\PostRector\Rector\NameImportingPostRector
+ */
+final class AutoImportNamesOnlyOnChangedFiles extends AbstractRectorTestCase
```
Comment (samsonasik) [inline rules-tests/CodingStyle/Rector/Namespace_/ImportFullyQualifiedNamesRector/AutoImportNamesOnlyOnChangedFiles.php]: should has Test suffix to make it runnable from directory:

```suggestion
final class AutoImportNamesOnlyOnChangedFilesTest extends AbstractRectorTestCase
```

Comment (sabbelasichon) [reply rules-tests/CodingStyle/Rector/Namespace_/ImportFullyQualifiedNamesRector/AutoImportNamesOnlyOnChangedFiles.php]: How could i forget?

**Review by samsonasik — APPROVED** [review]

**Comment by sabbelasichon** [comment]
@TomasVotruba For the first time, i would say this is enough. In the future we can create separate rector just for the AutoImportNames stuff. So ready to merge ;-)

File: packages/PostRector/Rector/NameImportingPostRector.php [hunk 2]
```diff
@@ -128,4 +137,13 @@ private function shouldImportName(Name $name, File $file, array $currentUses): b
...
         return $this->reflectionProvider->hasFunction(new Name($name->getLast()), null);
     }
+
+    private function shouldApply(File $file): bool
+    {
+        if ($file->hasContentChanged()) {
+            return true;
+        }
+
+        return ! $this->parameterProvider->provideBoolParameter(Option::APPLY_AUTO_IMPORT_NAMES_ON_CHANGED_FILES_ONLY);
```
Comment (samsonasik) [inline packages/PostRector/Rector/NameImportingPostRector.php]: I think comparing content can be only when there is `APPLY_AUTO_IMPORT_NAMES_ON_CHANGED_FILES_ONLY` enabled:

```suggestion
        if (! (bool) $this->parameterProvider->provideBoolParameter(Option::APPLY_AUTO_IMPORT_NAMES_ON_CHANGED_FILES_ONLY)) {
            return true;
        }

        return $file->hasContentChanged();
```

**Comment by TomasVotruba** [comment]
I would like to merge it.

I wonder why CI is failing?

**Comment by samsonasik** [comment]
that possibly needs rebase

**Comment by TomasVotruba** [comment]
Could you check?

**Comment by samsonasik** [comment]
rebased.

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
I am merging it ;)

**Comment by samsonasik** [comment]
Thank you @sabbelasichon

**Comment by TomasVotruba** [comment]
Awesome, thank you both :+1:

**Comment by sabbelasichon** [comment]
@samsonasik Thank you very much.

**Pull request merged**